### ISSUE TITLE
Obscure error when login shell does not understand `exit 71;`

### ISSUE THREAD

**Issue by tiffany352** [cockpit-project/cockpit#14687]
Cockpit version: 228
OS: Fedora 32 Server
Page: Login

Trying to login to cockpit gives me a 403 forbidden error from `/cockpit/login`. I had this setup previously working, but when I changed my default shell a few weeks ago I broke it without realizing.

After much frustrated debugging, I narrowed it down to this:
```
Sep 30 20:50:32 home.tiffnix.com cockpit-ws[10677]: error: exit unexpected 71
Sep 30 20:50:32 home.tiffnix.com cockpit-ws[10677]:   ┌─ shell:1:6
Sep 30 20:50:32 home.tiffnix.com cockpit-ws[10677]:   │
Sep 30 20:50:32 home.tiffnix.com cockpit-ws[10677]: 1 │ exit 71;
Sep 30 20:50:32 home.tiffnix.com cockpit-ws[10677]:   │      ^^ unexpected argument (try exit -h)
```
Corresponding to this line of code that's supposed to check if the user has a valid login shell:
https://github.com/cockpit-project/cockpit/blob/dd3a40419d7fb06b3479afde89017d9111fe91bc/src/ws/session-utils.c#L876-L883
https://github.com/cockpit-project/cockpit/blob/dd3a40419d7fb06b3479afde89017d9111fe91bc/src/ws/session.c#L711-L712

## Steps to reproduce

1. Install NuShell or any other shell where `exit 71` isn't a valid command.
2. `chsh` to it.
3. Try to login to cockpit.

## Screenshots
![image](https://user-images.githubusercontent.com/1254344/94766718-82bb7880-0360-11eb-9e39-92a53c3f43a9.png)

![image](https://user-images.githubusercontent.com/1254344/94764889-d9748280-035f-11eb-8e2c-0bd6c6e5b71d.png)

## Additional info

PAM was returning success codes, which made this extra confusing. It was only once I looked at `journalctl -f` output that I realized what was happening.

I get that cockpit needs to do this check, but it would be nice if the error was surfaced better, it took a lot of digging to figure this out.

I did realize after this that maybe I should enter into nushell from bashrc or something instead of using `chsh`.

**Comment by marusak** [comment]
@allisonkarlitskaya do you think it makes sense to replace `PAM_PERM_DENIED` with some different error code that would directly indicate that login shell is not valid?

**Comment by allisonkarlitskaya** [comment]
@tiffany352 is the login shell listed in `/etc/shells`?

I think maybe we need to up our game here... again.

**Comment by tiffany352** [comment]
> @tiffany352 is the login shell listed in `/etc/shells`?
> 
> I think maybe we need to up our game here... again.

I think I added it there manually. I install Nu via this Copr repo for Fedora: https://copr.fedorainfracloud.org/coprs/atim/nushell/

`chsh` won't even let you change the shell to a shell that isn't listed there, if I remember correctly. So it might work as a way of detecting a shell's validity, assuming there's no distro that lists `nologin` there.

The main reason I filed this issue was just the error message though - if it said your login shell was invalid I'd have immediately known what was up. That would help out anyone else who runs into this, like Nu users or other bespoke shells.

For completeness, I worked around this issue by setting my shell to bash and adding this to my bashrc:
```bash
if [ ! -z "$PS1" ]; then
        exec nu
fi
```
